query/parser.go

```diff
@@ -99,57 +99,55 @@ var dateFormats = []string{
 	time.RFC822Z,
 }
 
-// `now` isn't captured in this regex
-var relativeTimeRegexp = regexp.MustCompile(`^-(\d+)([a-zA-Z])$`)
+var relativeTimeRegexp = regexp.MustCompile(`^now\s*([+-])\s*([0-9]+)([smhdwMy])$`)
 
-// parseRelativeTime transforms a relative time string (e.g. `-10m`) to a
-// timestamp in millis.
-func parseRelativeTime(s string, now time.Time) (int64, error) {
-	if strings.ToLower(s) == "now" {
-		return now.Unix() * 1000, nil
+// parseDate converts the given datestring (from one of the allowable formats) into a millisecond offset from the Unix epoch.
+func parseDate(date string) (int64, error) {
+
+	if date == "now" {
+		return time.Now().Unix() * 1000, nil
 	}
 
-	matches := relativeTimeRegexp.FindStringSubmatch(s)
+	matches := relativeTimeRegexp.FindStringSubmatch(date)
 	if matches != nil {
-		offsetSeconds, err := strconv.ParseInt(matches[1], 10, 0)
+		// This match can be used to determine the duration of time.
+		now := time.Now().Unix()
+		sign := int64(1)
+		if matches[1] == "-" {
+			sign = -1
+		}
+		offset, err := strconv.ParseInt(matches[2], 10, 0)
 		if err != nil {
-			// This should never happen since the regex should be valid
-			panic(fmt.Sprintf("Invalid offset `%s` in relative timestamp `%s`", matches[1], s))
+			// This case could occur when the number is too large, for example.
+			return 0, err
 		}
-
-		offset := offsetSeconds * 1000
-		specifier := matches[2]
-
-		switch specifier {
+		switch matches[3] {
 		case "s":
-			// Already in millis, do nothing.
+			offset *= 1
 		case "m":
 			offset *= 60
 		case "h":
 			offset *= 60 * 60
 		case "d":
 			offset *= 60 * 60 * 24
+		case "w":
+			offset *= 60 * 60 * 24 * 7
 		case "M":
 			offset *= 60 * 60 * 24 * 30
 		case "y":
 			offset *= 60 * 60 * 24 * 365
 		default:
-			return -1, fmt.Errorf("invalid relative timestamp %s", s)
+			// Won't happen, regex can't produce any other value.
+			panic(matches[3])
```

Comment:
The code is unreachable, provided the timestamp regexp (and the switch) is correct/consistent, so I decided to panic.